tests/decomposition/test_decomposition_graph.py

```diff
@@ -42,6 +42,57 @@
 )
 class TestDecompositionGraph:
 
+    def test_weighted_graph_solve(self, _):
+        """Tests solving a simple graph for the optimal decompositions with weighted gates."""
+
+        op = qml.CRX(2.5, wires=[0, 1])
+
+        # the RZ CZ RX CZ decomp is chosen when the RZ and CNOT weights are large.
+        graph = DecompositionGraph(
+            operations=[op],
+            gate_set={"RX": 1.0, "RY": 3.0, "RZ": 10.0, "GlobalPhase": 1.0, "CNOT": 20.0, "CZ": 1.0},
+        )
+        graph.solve()
+
+        expected_resource = to_resources({qml.CZ: 2, qml.RX: 2})
+        assert graph.resource_estimate(op) == expected_resource
+
+        # the RZ CZ RX CZ decomp is avoided when the CZ weight is large.
+        graph = DecompositionGraph(
+            operations=[op],
+            gate_set={"RX": 1.0, "RY": 1.0, "RZ": 1.0, "GlobalPhase": 1.0, "CNOT": 1.0, "CZ": 10.0},
+        )
+        graph.solve()
+
+        expected_resource = to_resources({qml.RX: 2, qml.CNOT: 2, qml.RY: 4, qml.GlobalPhase: 4, qml.RZ: 4})
+        assert graph.resource_estimate(op) == expected_resource
+
+        # bad input: string
+        with pytest.raises(TypeError):
+            graph = DecompositionGraph(
+                operations=[op],
+                gate_set={"RX": "1.0", "RY": 1.0, "RZ": 1.0, "GlobalPhase": 1.0, "CNOT": 1.0, "CZ": 10.0},
+            )
+            graph.solve()
+
+        # edge case: negative gate weight
+        with pytest.raises(ValueError, match="Negative weights not supported."):
+            graph = DecompositionGraph(
+                operations=[op],
+                gate_set={"RX": 1.0, "RY": 1.0, "RZ": 1.0, "GlobalPhase": 1.0, "CNOT": 1.0, "CZ": -10.0},
+            )
+            graph.solve()
```

Comment:
In general, we want each test to have a single responsibility. For example, these should be in one or even two separate test cases.